### PR TITLE
devex: initial commit of devcontainer setup

### PR DESCRIPTION
**Pull request by lizan**
Risk Level: Low (not affecting any output)
Testing: local
Docs Changes: `.devcontainer/README.md`
Release Notes: N/A

Signed-off-by: Lizan Zhou <lizan@tetrate.io>

### CODE REVIEW

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,21 @@
+FROM gcr.io/envoy-ci/envoy-build:04f06115b6ee7cfea74930353fb47a41149cbec3
+
+ARG USERNAME=vscode
+ARG USER_UID=501
+ARG USER_GID=$USER_UID
+
+ENV BUILD_DIR=/build
```
Comment (htuch) [inline .devcontainer/Dockerfile:7]: For whatever reason, when working with this on Linux, I needed to chmod this to be a+w (or alternatively put in vscode owner/group)

File: .devcontainer/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,21 @@
+FROM gcr.io/envoy-ci/envoy-build:04f06115b6ee7cfea74930353fb47a41149cbec3
+
+ARG USERNAME=vscode
```
Comment (htuch) [inline .devcontainer/Dockerfile:3]: I think what is here to start with is fine. I think maybe add a TODO around migrating this to the actual UID outside the container (which is how `run_envoy_docker.sh` works and matches Linux ownership semantics if you agree.

File: .devcontainer/README.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
+# Envoy Dev Container (experimental)
+
+This directory contains some experimental tools for Envoy Development in [VSCode Remote - Containers](https://code.visualstudio.com/docs/remote/containers).
+
+## How to use
+
+Open with VSCode with the Container extension installed. Follow the [official guide](https://code.visualstudio.com/docs/remote/containers) to open this
```
Comment (htuch) [inline .devcontainer/README.md:7]: Can you be a bit more precise in wording here. I think it's the `Remote - Containers` extension.

File: .devcontainer/README.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+### Using Remote Build Execution
+
+Write the following content to `devcontainer.env` and rebuild the container. The key will be persisted in the container's `~/.bazelrc`.
+
+```
+GCP_SERVICE_ACCOUNT_KEY=<base64 encoded service account key>
+BAZEL_REMOTE_INSTANCE=<RBE Instance>
+BAZEL_REMOTE_CACHE=grpcs://remotebuildexecution.googleapis.com
+BAZEL_BUILD_EXTRA_OPTIONS=--config=remote-ci --config=remote --jobs=<Number of jobs>
```
Comment (htuch) [inline .devcontainer/README.md:24]: Is it worth diving into download minimal vs. all here? I think that can make a huge difference if you are sitting behind a cable modem connection vs. same region GCP?

Comment (lizan) [reply .devcontainer/README.md:24]: added.

File: .devcontainer/Dockerfile (hunk 1 above)
Comment (lizan) [reply .devcontainer/Dockerfile:7]: added in setup.sh to do chown

File: .devcontainer/Dockerfile (hunk 2 above)
Comment (lizan) [reply .devcontainer/Dockerfile:3]: I added remoteUser/containerUser in devcontainer.json, so I think VSCode will update this?

File: .devcontainer/README.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+```
+
+By default the `--config=remote` implies [`--remote_download_toplevel`](https://docs.bazel.build/versions/master/command-line-reference.html#flag--remote_download_toplevel),
+change this to `minimal` or `all` depending on where you're running the container by adding them to `BAZEL_BUILD_EXTRA_OPTIONS`.
+
+### Disk performance
+
+Docker for Mac/Windows is known to have disk performance issue, this makes formatting all files in the container very slow.
+[Update the mount consistency to 'delegated'](https://code.visualstudio.com/docs/remote/containers-advanced#_update-the-mount-consistency-to-delegated-for-macos) is recommended.
```
[the remark this answers is not in the document]
Comment (lizan) [reply .devcontainer/README.md:33]: This goes to user's `settings.json` so I don't have any way to enable it by default.

[the remark this answers is not in the document]
Comment (snowp) [reply .devcontainer/README.md:35]: There is a `workspaceMount` option you can set on devcontainer.json, but maybe we don't want to mess with the mount options by default.

File: .devcontainer/Dockerfile (hunk 2 above)
Comment (htuch) [reply .devcontainer/Dockerfile:3]: It didn't seem to when I did my last experiment with this, but that was remote Docker on a different host, so maybe that was stretching it. Let's go with this for now and update as we go.

File: tools/vscode/refresh_compdb.sh [hunk 6]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+
+tools/proto_format/proto_format.sh fix
+
+# Setting platform suffix here so the compdb headers won't be overwritten by another bazel run
```
Comment (htuch) [inline tools/vscode/refresh_compdb.sh:5]: As discussion on Slack, probably worth some warnings on what needs to be done when doing API work in this environment.

Comment (lizan) [reply tools/vscode/refresh_compdb.sh:5]: Added in README.md https://github.com/envoyproxy/envoy/pull/11207/files#diff-00554144415ae349ed4a355a828dd494R14

File: .devcontainer/README.md (hunk 3 above)
Comment (lizan) [reply .devcontainer/README.md:7]: That's in L3?

**Pull request merged**